### ISSUE TITLE
Avoid error message regarding cacerts when running with read-only file system

### ISSUE THREAD

**Issue by schnatterer** [paketo-buildpacks/libjvm#6]
Running containers with read-only file system increases runtime security.

While this is possible with images created by this build pack, and error message like the following is logged on container startup:

```
  E 2020-07-17T12:34:01.498115853Z unable to open /layers/paketo-buildpacks_bellsoft-liberica/jre/lib/security/cacerts
  E 2020-07-17T12:34:01.498122355Z open /layers/paketo-buildpacks_bellsoft-liberica/jre/lib/security/cacerts: read-only file system
```

* Are there any consequences when the `cacerts` file is not being modified as intended?
* Can this be avoided or the message deactivated?

**Comment by schnatterer** [comment]
@nebhale could you elaborate on the functional deficiency? If I don't intend to add certs, what deficits are there?
I have been running workloads with read-only file system for quite some time now.

It is challenging for some images but provides a lot of extra security by making the application code and configuration immutable.

**Comment by nebhale** [comment]
The two things that come to mind are:

1. adding certs, a _very_ common behavior in corporate enterprise environments (company-wide CA, not signed by a global CA)
2. identifying local DNS servers and disabling JVM-local DNS caching, because Java predates DNS TTL, it caches for fixed periods ("always" by default) making very-dynamic DNSs like Kubernetes more-or-less broken

We might have other things (I need to inventory everything we do), but those two are top-of-mind.

**Comment by schnatterer** [comment]
I usually work around number one by creating and mounting a `cacerts` file that contains the required CAs.
For number two - in what way does a RO FS cause problems with DNS resolution?

What causes the error message? Is the `cacerts` file modified at startup?
As the error does not kill the container - would it be possible to disable the error message?

I've been running multiple containers build using the bellsoft liberica build pack with RO FS just fine for some weeks.
But I'm concerned about this error message on container startup. What consequences are there when `libjvm` can't open `cacerts`?

**Comment by nebhale** [comment]
Because Java, to modify DNS properties, you must [write two values](https://github.com/paketo-buildpacks/libjvm/blob/main/dns/link_local_dns.go#L45-L48) into the file identified by `-Djava.security.properties` (or the default in the JVM). 😢 

For a little background on this, the following [networking properties](https://docs.oracle.com/javase/7/docs/technotes/guides/net/properties.html) are of interest:

> ### Address Cache
>
> The java.net package, when doing name resolution, uses an address cache for both security and performance reasons. Any address resolution attempt, be it forward (name to IP address) or reverse (IP address to name), will have its result cached, whether it was successful or not, so that subsequent identical requests will not have to access the naming service. These properties allow for some tuning on how the cache is operating.
>
> **networkaddress.cache.ttl** (default: see below)
Value is an integer corresponding to the number of seconds successful name lookups will be kept in the cache. A value of -1, or any other negative value for that matter, indicates a “cache forever” policy, while a value of 0 (zero) means no caching. The default value is -1 (forever) if a security manager is installed, and implementation specific when no security manager is installed.
>
> **networkaddress.cache.negative.ttl** (default: 10)
Value is an integer corresponding to the number of seconds an unsuccessful name lookup will be kept in the cache. A value of -1, or any negative value, means “cache forever”, while a value of 0 (zero) means no caching.
>
> Since these 2 properties are part of the security policy, they are not set by either the -D option or the System.setProperty() API, instead they are set as security properties.

The root of the problem is that in a system like Kubernetes, the DNS content is *very* dynamic (every time a pod comes and goes, the IPs backing a DNS entry are updated).  If you have to wait 10s, 5s, hell even 1s (and remember, it's implementation specific when unspecified, so who the hell knows how long the delay is...), then as pods are updated it appears as downtime to callers.  In such a situation, the correct behavior is to not cache at all (especially if you can't honor the TTLs that Kubernetes passes you) and pray that bring on the same cluster as the DNS makes your DNS calls near-instant.

Now where this gets interesting is that when a CNB is building an image **it has no idea where that image will run**.  It _could_ run in Kubernetes where making this change makes sense.  It might also run on some DIY Docker-based cloud in a datacenter where this change would lead to catastrophic performance problems.  This gets us to trying to work it out, at launch time.

---

CA certs are a similar problem.  When a CNB builds an image, it knows what certificates are available to it **in the build image**.  To start with, there's no guarantee that a [build and run image](https://buildpacks.io/docs/concepts/components/stack/) have the same set of certs since they are different images, [built](https://github.com/paketo-buildpacks/stacks/blob/main/bionic/dockerfile/build/Dockerfile) at [different](https://github.com/paketo-buildpacks/stacks/blob/main/bionic/dockerfile/run/Dockerfile) times.

Compounding this, CNB has an awesome feature called [rebasing](https://buildpacks.io/docs/concepts/operations/rebase/) that allows you to swap the Operating System underneath an application without touching the upper layers (say the one with the JVM's `cacerts` in it).  Given that one of the things that is almost always included in, and updated by, stacks is the [`ca-certificates` package](https://github.com/paketo-buildpacks/stacks/blob/main/tiny/dockerfile/run/packagelist#L2), the CA certs collection is one of the things most likely to change during a rebase.  So this again, leads us to trying to convert OpenSSL-based CA certs at the time we know the absolute truth, launch time.

---

Now all of this is **MY** problem not yours (😇) and we do want to support your RO filesystem needs.  To that end we're going to do the following:

1. At build time we're going to update the JRE `cacerts` with the certificates that are available in the build image.  For people who want to run Java applications in RO, this means they'll have to keep the build and run images in sync cert-wise.
1. Update all utilities that attempt to modify the filesystem at launch time, to gracefully handle RO filesystems and warn that they were not able to do the work that they wanted to.

These changes mean that you'll be stuck with the sub-standard DNS behavior (in cases where you might not otherwise be), and won't be able to use rebase to update CA certs (you'll have to do a multi-second/minute build versus a 10ms rebase to get new certs).  Obviously, neither of these are deal breakers, just heart-breakers that I can't make your applications as awesome as they otherwise might be 😜.

P.S. You did find an actual bug in the buildpack though.  When the `cacerts` couldn't be written the entire application should have refused to start at all.  I'm glad that bug has allowed you to proceed this far, but clearly we have to rethink how to ensure that we're failing at the appropriate times.

**Comment by schnatterer** [comment]
Thanks for the enlightening details, @nebhale!
Once you start digging into CNB the learning curve gets steeper, but I'm slowly creeping upwards :smile: 

Those changes sound very sensible to me! A detailed warning is much better to handle than an error.

So for my use case, I think I'm fine with CAs. Sorry for the heart breaking :wink: 
I learned that I should improve JVM DNS caching, though. In order to disable JVM DNS caching for my k8s deployments, I'm considering mounting a `security.properties` file containing the DNS TTL options into my container. Do you think this would work?
I can offer to document my findings in some appropriate place (where?), as I can image they are interesting for other security-sensitive folks.

**Comment by nebhale** [comment]
Well, learning curve perhaps.  We like to think of it as doing a bunch of best-practice work so that you never need to know that it's happening.

Mounting your own `security.properties` in place will likely work.

**Comment by nebhale** [comment]
Resolved by 20962c26e6c177a27889dff4b813ab1f8ab990af
